File: js/Core/Helper/Collections.js
import { disposeSafe, getEnumerator, defaultOf, structuralHash, equals } from "../../fable_modules/fable-library-js.4.22.0/Util.js";
import { value as value_1, some } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { empty, singleton, append, head, tail, isEmpty } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { Dictionary } from "../../fable_modules/fable-library-js.4.22.0/MutableMap.js";
import { iterate } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";
import { tryGetValue, addToDict } from "../../fable_modules/fable-library-js.4.22.0/MapUtil.js";
import { FSharpRef } from "../../fable_modules/fable-library-js.4.22.0/Types.js";
import { contains } from "../../fable_modules/fable-library-js.4.22.0/Array.js";

/**
 * If the value matches the default, a None is returned, else a Some is returned
 */
export function Option_fromValueWithDefault(d, v) {
    if (equals(d, v)) {
        return undefined;
    }
    else {
        return some(v);
    }
}

/**
 * Applies the function f on the value of the option if it exists, else applies it on the default value. If the result value matches the default, a None is returned
 */
export function Option_mapDefault(d, f, o) {
    return Option_fromValueWithDefault(d, (o == null) ? f(d) : f(value_1(o)));
}

/**
 * Applies the function f on the value of the option if it exists, else returns the default value.
 */
export function Option_mapOrDefault(d, f, o) {
    if (o == null) {
        return d;
    }
    else {
        return some(f(value_1(o)));
    }
}

export function List_tryPickAndRemove(f, lst) {
    const loop = (newList_mut, remainingList_mut) => {
        loop:
        while (true) {
            const newList = newList_mut, remainingList = remainingList_mut;
            if (!isEmpty(remainingList)) {
                const t = tail(remainingList);
                const h = head(remainingList);
                const matchValue = f(h);
                if (matchValue == null) {
                    newList_mut = append(newList, singleton(h));
                    remainingList_mut = t;
                    continue loop;
                }
                else {
                    return [some(value_1(matchValue)), append(newList, t)];
                }
            }
            else {
                return [undefined, newList];
            }
            break;
        }
    };
    return loop(empty(), lst);
}

export function Dictionary_ofSeq(s) {
    const dict = new Dictionary([], {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    iterate((tupledArg) => {
        addToDict(dict, tupledArg[0], tupledArg[1]);
    }, s);
    return dict;
}

export function Dictionary_tryFind(key, dict) {
    let patternInput;
    let outArg = defaultOf();
    patternInput = [tryGetValue(dict, key, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (patternInput[0]) {
        return some(patternInput[1]);
    }
    else {
        return undefined;
    }
}

export function Dictionary_ofSeqWithMerge(merge, s) {
    const dict = new Dictionary([], {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    iterate((tupledArg) => {
        const k = tupledArg[0];
        const v = tupledArg[1];
        const matchValue = Dictionary_tryFind(k, dict);
        if (matchValue == null) {
            addToDict(dict, k, v);
        }
        else {
            const v$0027 = value_1(matchValue);
            dict.delete(k);
            addToDict(dict, k, merge(v$0027, v));
        }
    }, s);
    return dict;
}

export function ResizeArray_map(f, a) {
    const b = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            void (b.push(f(i)));
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return b;
}

export function ResizeArray_choose(f, a) {
    const b = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const matchValue = f(enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]());
            if (matchValue == null) {
            }
            else {
                const x = value_1(matchValue);
                void (b.push(x));
            }
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return b;
}

export function ResizeArray_filter(f, a) {
    const b = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            if (f(i)) {
                void (b.push(i));
            }
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return b;
}

export function ResizeArray_fold(f, s, a) {
    let state = s;
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            state = f(state, i);
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return state;
}

export function ResizeArray_foldBack(f, a, s) {
    let state = s;
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            state = f(i, state);
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return state;
}

export function ResizeArray_iter(f, a) {
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            f(enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]());
        }
    }
    finally {
        disposeSafe(enumerator);
    }
}

export function ResizeArray_reduce(f, a) {
    switch (a.length) {
        case 0:
            throw new Error("ResizeArray.reduce: empty array");
        case 1:
            return a[0];
        default: {
            const a_5 = a;
            let state = a_5[0];
            for (let i = 1; i <= (a_5.length - 1); i++) {
                state = f(state, a_5[i]);
            }
            return state;
        }
    }
}

export function ResizeArray_collect(f, a) {
    const b = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const enumerator_1 = getEnumerator(f(enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]()));
            try {
                while (enumerator_1["System.Collections.IEnumerator.MoveNext"]()) {
                    const j = enumerator_1["System.Collections.Generic.IEnumerator`1.get_Current"]();
                    void (b.push(j));
                }
            }
            finally {
                disposeSafe(enumerator_1);
            }
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return b;
}

export function ResizeArray_distinct(a) {
    const b = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            if (!contains(i, b, {
                Equals: equals,
                GetHashCode: structuralHash,
            })) {
                void (b.push(i));
            }
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return b;
}

export function ResizeArray_isEmpty(a) {
    return a.length === 0;
}

/**
 * Immutable append
 */
export function ResizeArray_append(a, b) {
    const c = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            void (c.push(i));
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    let enumerator_1 = getEnumerator(b);
    try {
        while (enumerator_1["System.Collections.IEnumerator.MoveNext"]()) {
            const i_1 = enumerator_1["System.Collections.Generic.IEnumerator`1.get_Current"]();
            void (c.push(i_1));
        }
    }
    finally {
        disposeSafe(enumerator_1);
    }
    return c;
}

/**
 * append a single element
 */
export function ResizeArray_appendSingleton(b, a) {
    const c = [];
    let enumerator = getEnumerator(a);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            void (c.push(i));
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    void (c.push(b));
    return c;
}

File: src/Core/Helper/Collections.fs
namespace ARCtrl.Helper

open Fable.Core

module Seq =
    let inline compare (a: seq<'a>) (b: seq<'a>) =
        if Seq.length a = Seq.length b then
            [ for i in 0 .. (Seq.length a) - 1 do yield (Seq.item i a = Seq.item i b) ]
            |> Seq.fold (fun acc e -> acc && e) true
        else
            false

module Option =

    /// If the value matches the default, a None is returned, else a Some is returned
    let fromValueWithDefault d v =
        if d = v then None
        else Some v

    /// Applies the function f on the value of the option if it exists, else applies it on the default value. If the result value matches the default, a None is returned
    let mapDefault (d : 'T) (f: 'T -> 'T) (o : 'T option) =
        match o with
        | Some v -> f v
        | None   -> f d
        |> fromValueWithDefault d

    /// Applies the function f on the value of the option if it exists, else returns the default value.
    let mapOrDefault (d : 'T Option) (f: 'U -> 'T) (o : 'U option) =
        match o with
        | Some v -> Some (f v)
        | None   -> d

    /// If the value matches the default, a None is returned, else a Some is returned
    let fromSeq (v : 'T when 'T :> System.Collections.IEnumerable) =
        if Seq.isEmpty v then None
        else Some v

module internal List =

    let tryPickAndRemove (f : 'T -> 'U option) (lst : 'T list) =
        let rec loop newList remainingList =
            match remainingList with
            | h::t ->
                match f h with
                | Some v -> Some v, newList @ t
                | None -> loop (newList @ [h]) t
            | _ -> None, newList
        loop [] lst

module Dictionary =

    open System.Collections.Generic

    let addOrUpdate (key : 'Key) (value : 'T) (dict : Dictionary<'Key,'T>) =
        if dict.ContainsKey key then
            dict.[key] <- value
        else
            dict.Add(key,value)

    let ofSeq (s : seq<'Key*'T>) =
        let dict = Dictionary()
        s
        |> Seq.iter dict.Add
        dict

    let tryFind (key : 'Key) (dict : Dictionary<'Key,'T>) =
        let b,v = dict.TryGetValue key
        if b then Some v
        else None

    let ofSeqWithMerge (merge : 'T -> 'T -> 'T) (s : seq<'Key*'T>) =
        let dict = Dictionary()
        s
        |> Seq.iter (fun (k,v) ->
            match tryFind k dict with
            | Some v' ->
                dict.Remove(k) |> ignore
                dict.Add(k,merge v' v)
            | None ->
                dict.Add(k,v)
            )
        dict

    let init () : Dictionary<'Key,'T> =
        #if FABLE_COMPILER_PYTHON
        Fable.Core.PyInterop.emitPyExpr () "dict()"
        #else
        Dictionary<'Key,'T>()
        #endif

    let items (dict : Dictionary<'Key,'T>) : seq<KeyValuePair<'Key, 'T>> =
        #if FABLE_COMPILER_PYTHON
        Fable.Core.PyInterop.emitPyExpr (dict) "$0.items()"
        #else
        dict
        #endif

module StringDictionary =

    open System.Collections.Generic

    let ofSeq (s : seq<string*string>) : Dictionary<string,string> =
        s
        |> dict
        #if !FABLE_COMPILER
        |> Dictionary
        #else
        |> unbox
        #endif

    let inline tryFind (key : string) (dict : Dictionary<string,'T>) =
        let b,v = dict.TryGetValue key
        if b then Some v
        else None

    let inline addOrUpdate (key : string) (value : 'T) (dict : Dictionary<string,'T>) =
        if dict.ContainsKey key then
            dict.[key] <- value
        else
            dict.Add(key,value)


module ResizeArray =

    open System.Collections.Generic

    let create (i : int) (v : 'T) =
        let a = ResizeArray<_>()
        if i > 0 then
            for _ in 1 .. i do
                a.Add(v)
        a

    let singleton (a : 'T) =
        let b = ResizeArray<_>()
        b.Add(a)
        b

    let map  f (a : ResizeArray<_>) =
        let b = ResizeArray<_>()
        for i in a do
            b.Add(f i)
        b

    let choose f (a : ResizeArray<_>) =
        let b = ResizeArray<_>()
        for i in a do
            match f i with
            | Some x -> b.Add(x)
            | None -> ()
        b

    let filter f (a : ResizeArray<_>) =
        let b = ResizeArray<_>()
        for i in a do
            if f i then b.Add(i)
        b

    let fold f s (a : ResizeArray<_>) =
        let mutable state = s
        for i in a do
            state <- f state i
        state

    let foldBack f (a : ResizeArray<_>) s =
        let mutable state = s
        for i in a do
            state <- f i state
        state

    let iter f (a : ResizeArray<_>) =
        for i in a do
            f i

    let init (n : int) (f : int -> 'T) =
        let a = ResizeArray<_>()
        for i in 0 .. n - 1 do
            a.Add(f i)
        a


    let reduce f (a : ResizeArray<_>) =
        match a with
        | a when a.Count = 0 -> failwith "ResizeArray.reduce: empty array"
        | a when a.Count = 1 -> a.[0]
        | a ->
            let mutable state = a.[0]
            for i in 1 .. a.Count - 1 do
                state <- f state a.[i]
            state

    let collect f (a : ResizeArray<_>) =
        let b = ResizeArray<_>()
        for i in a do
            let c = f i
            for j in c do
                b.Add(j)
        b

    let distinct (a : ResizeArray<_>) =
        let b = ResizeArray<_>()
        for i in a do
            if not (b.Contains(i)) then
                b.Add(i)
        b

    let isEmpty (a : ResizeArray<_>) =
        a.Count = 0

    /// Immutable append
    let append (a : ResizeArray<_>) (b : ResizeArray<_>) =
        let c = ResizeArray<_>()
        for i in a do
            c.Add(i)
        for i in b do
            c.Add(i)
        c

    /// append a single element
    let appendSingleton (b : 'T) (a : ResizeArray<_>) =
        let c = ResizeArray<_>()
        for i in a do
            c.Add(i)
        c.Add(b)
        c


    // Make sure that output type matches
    let groupBy (f : 'T -> 'a) (a : ResizeArray<'T>) : ResizeArray<'a*ResizeArray<'T>> =
        Seq.groupBy f a
        |> Seq.map (fun (k,v) -> k, ResizeArray v)
        |> ResizeArray

    let tryPick f (a : ResizeArray<'T>) =
        let rec loop i =
            if i < a.Count then
                match f a.[i] with
                | Some v -> Some v
                | None -> loop (i + 1)
            else None
        loop 0

    let zip (a : ResizeArray<'T>) (b : ResizeArray<'U>) =
        let c = ResizeArray<_>()
        let n = min a.Count b.Count
        for i in 0 .. n - 1 do
            c.Add(a.[i], b.[i])
        c

    let tryFind f (a : ResizeArray<'T>) =
        let rec loop i =
            if i < a.Count then
                if f a.[i] then
                    Some a.[i]
                else
                   loop (i + 1)
            else None
        loop 0